runtime/health.go

```diff
@@ -57,6 +52,163 @@ func (r *Runtime) Health(ctx context.Context) (*Health, error) {
 	}, nil
 }
 
-func (r *Runtime) InstanceHealth(ctx context.Context, instanceID string) (*InstanceHealth, error) {
-	return r.registryCache.instanceHealth(ctx, instanceID)
+func (r *Runtime) InstanceHealth(ctx context.Context, instanceID string, query DashboardHealthQuery) (*InstanceHealth, error) {
+	res := &InstanceHealth{}
+	// check repo error
+	repo, rr, err := r.Repo(ctx, instanceID)
+	if err != nil {
+		res.Repo = err.Error()
+	} else {
+		err = repo.(drivers.Handle).Ping(ctx)
+		if err != nil {
+			res.Repo = err.Error()
+		}
+		rr()
+	}
+
+	ctrl, err := r.Controller(ctx, instanceID)
+	if err != nil {
+		res.Controller = err.Error()
+		return res, nil
+	}
+
+	// check olap error
+	olap, release, err := r.OLAP(ctx, ctrl.InstanceID, "")
+	if err != nil {
+		res.OLAP = err.Error()
+		return res, nil
+	}
+	defer release()
+
+	resources, err := ctrl.List(ctx, ResourceKindMetricsView, "", false)
+	if err != nil {
+		return nil, err
+	}
+
+	cacheEnabled := healthCheckCacheEnabled(olap)
+	if cacheEnabled {
+		instanceHealth, ok := r.cachedInstanceHealth(ctx, ctrl.InstanceID, ctrl.catalog.version, resources)
+		if ok {
+			res.OLAP = instanceHealth.OLAP
+			res.Dashboards = instanceHealth.Dashboards
+			return res, nil
+		}
+	}
+
+	err = olap.(drivers.Handle).Ping(ctx)
+	if err != nil {
+		res.OLAP = err.Error()
+	} else if query != nil {
+		// run queries against dashboards
+		res.Dashboards = make(map[string]string, len(resources))
+		for _, mv := range resources {
+			q, err := query(ctx, ctrl.InstanceID, mv.Meta.Name.Name)
```

Comment:
I was also split on this. 
A few times I have seen requests about current health check returning healthy for cases when dashboards aren't healthy so that's why proceeded to check on invalid metrics views.
May be let's throw errors for all to begin with and if we get complaints about too many alerts on user errors we can change this ?

Reply:
But if the metrics view is invalid (doesn't have `ValidSpec`), the query is guaranteed to fail.

I think parse/reconcile errors are a different case from queries to a valid metrics view not responding to queries, but I agree parse/reconcile errors are also useful to track.

What about if we skip invalid metrics views from the queries here, and instead add `uint parse_errors` and `uint reconcile_errors` to the `InstanceHealth` for reporting if and how many of such errors the instance has? Then we could optionally also monitor for that.